src/xfastertransformer/tools/llama_convert.py

```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
+            config["llama"]["weight_data_type"] = dtype
+
+            config["llama"]["quant_decoder_weights"] = str(True)
+            self.wbits = quantize_config["bits"]
+            assert self.wbits == 8 or self.wbits == 4, "Only 4/8bits quantization is supported"
+            config["llama"]["quant_wbits"] = str(self.wbits)
+            assert quantize_config["group_size"] == -1, "Only column wise quantization is supported."
+            config["llama"]["quant_groupsize"] = str(quantize_config["group_size"])
+            #config["llama"]["quant_scheme"] = "sym" if quantize_config["sym"] == True else "asym"
+
+            with open(os.path.join(output_dir, "config.ini"), "w") as configfile:
+                config.write(configfile)
+        except Exception as e:
+            print("Fail to save the config in config.ini.", str(e))
+
+        hf_model_name_pattern = [
+            "input_layernorm.weight",
+            "self_attn.qkv_proj.qweight",
+            "self_attn.qkv_proj.qzeros",
+            "self_attn.qkv_proj.scales",
+            "self_attn.o_proj.qweight",
+            "self_attn.o_proj.qzeros",
+            "self_attn.o_proj.scales",
+            "post_attention_layernorm.weight",
+            "mlp.gate_proj.qweight",
+            "mlp.gate_proj.qzeros",
+            "mlp.gate_proj.scales",
+            "mlp.up_proj.qweight",
+            "mlp.up_proj.qzeros",
+            "mlp.up_proj.scales",
+            "mlp.down_proj.qweight",
+            "mlp.down_proj.qzeros",
+            "mlp.down_proj.scales",
+        ]
+
+        ft_model_name_pattern = [
+            "input_layernorm.weight",
+            "attention.query_key_value.qweight",
+            "attention.query_key_value.zeros",
+            "attention.query_key_value.scales",
+            "attention.dense.qweight",
+            "attention.dense.zeros",
+            "attention.dense.scales",
+            "post_attention_layernorm.weight",
+            "mlp.gate_proj.qweight",
+            "mlp.gate_proj.zeros",
+            "mlp.gate_proj.scales",
+            "mlp.up_proj.qweight",
+            "mlp.up_proj.zeros",
+            "mlp.up_proj.scales",
+            "mlp.down_proj.qweight",
+            "mlp.down_proj.zeros",
+            "mlp.down_proj.scales",
+        ]
+
+        state_dict = model.state_dict()
+        model_named_parameters = dict()
+        for name, param in state_dict.items():
+            if name.startswith("model."):
+                name = name[6:]
+            wf = torch.tensor(list(range(0, 32, self.wbits)), dtype=torch.int32).unsqueeze(0)
+
+            print(name)
+
+            if "embed" in name:
+                model_named_parameters[name] = param
+            elif "lm_head" in name:
+                model_named_parameters[name] = param
+            elif "scales" in name:
+                model_named_parameters[name] = param.float()
+            elif "qzeros" in name:
+                qzeros = param
```

Comment:
zeros is uint8/uint4 from QPTQ quantization, and need to convert to fp32 for xFT.